generator.go

```diff
@@ -144,3 +149,88 @@ func sign(message, secret string) string {
 	h.Write([]byte(message))
 	return encode(h.Sum(nil))
 }
+
+func GenerateCustomToken(uid string, developerClaims *jws.Claims, clientEmail string, privateKeyString string) (string, error) {
+
+	privateKey, err := crypto.ParseRSAPrivateKeyFromPEM([]byte(privateKeyString))
+	if err != nil {
+		return "", err
+	}
+
+	if uid == "" {
+		return "", errors.New("Uid must be provided.")
+	}
+
+	if clientEmail == "" {
+		return "", errors.New("Must provide an issuer.")
+	}
+
+	method := crypto.SigningMethodRS256
+	claims := jws.Claims{}
+	claims.Set("uid", uid)
+	claims.SetIssuer(clientEmail)
+	claims.SetSubject(clientEmail)
+	claims.SetAudience(firebaseAudience)
+	now := time.Now()
+	claims.SetIssuedAt(now)
+	claims.SetExpiration(now.Add(time.Hour))
+
+	if developerClaims != nil {
+
+		for claim := range *developerClaims {
+			if isReserved(claim) {
+				return "", fmt.Errorf("developer_claims cannot contain a reserved key: %s", claim)
+			}
+		}
+		claims.Set("claims", developerClaims)
+	}
+
+	jwt := jws.NewJWT(claims, method)
+	bytes, err := jwt.Serialize(privateKey)
+	if err != nil {
+		return "", err
+	}
+
+	return string(bytes), nil
+}
+
+// isReserved determines whether a given name is a reserved name via binary search.
+
+var (
+	reservedNames = []string{
+		"acr",
+		"amr",
+		"at_hash",
+		"aud",
+		"auth_time",
+		"azp",
+		"cnf",
+		"c_hash",
+		"exp",
+		"firebase",
+		"iat",
+		"iss",
+		"jti",
+		"nbf",
+		"nonce",
+		"sub",
+	}
+)
+
+func isReserved(name string) bool {
+	if len(reservedNames) > 0 {
```

Comment:
You don't need this if statement. If the length of the array is 0 then `l` would be 0 and `r` would be -1 so the `for` loop would never execute.